plugin/opentelemetry/setup.go

```diff
@@ -0,0 +1,165 @@
+package opentelemetry
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+func init() { plugin.Register("opentelemetry", setup) }
+
+func setup(c *caddy.Controller) error {
+	ot, err := opentelemetryParse(c)
+	if err != nil {
+		return plugin.Error("opentelemetry", err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		ot.Next = next
+		return ot
+	})
+
+	c.OnStartup(func() error { return ot.OnStartup(c) })
+	c.OnFinalShutdown(func() error {
+		if ot.ShutdownFunc != nil {
+			return ot.ShutdownFunc(context.Background())
+		}
+		return nil
+	})
+
+	return nil
+}
+
+func opentelemetryParse(c *caddy.Controller) (*opentelemetry, error) {
+	var (
+		ot  = &opentelemetry{serviceName: defServiceName}
+		err error
+	)
+
+	for c.Next() {
+		var err error
```

Comment:
Creating `var err error` once should be enough here I think.

Reply:
you are right, fixed!